source/adaptive_load/adaptive_load_client_main.cc

```diff
@@ -0,0 +1,137 @@
+  const Envoy::Api::IoCallBoolResult open_result =
+      file->open(((1 << Envoy::Filesystem::File::Operation::Write)) |
+                 (1 << (Envoy::Filesystem::File::Operation::Create)));
+  if (!open_result.ok()) {
+    throw Nighthawk::NighthawkException(absl::StrCat("Unable to open output file \"", path,
+                                                     "\": ", open_result.err_->getErrorDetails()));
+  }
+  const Envoy::Api::IoCallSizeResult write_result = file->write(contents);
+  if (!write_result.ok()) {
+    throw Nighthawk::NighthawkException(absl::StrCat("Unable to write output file \"", path,
+                                                     "\": ", write_result.err_->getErrorDetails()));
+  }
+  const Envoy::Api::IoCallBoolResult close_result = file->close();
+  if (!close_result.ok()) {
+    throw Nighthawk::NighthawkException(absl::StrCat("Unable to close output file \"", path,
+                                                     "\": ", close_result.err_->getErrorDetails()));
+  }
+}
+
+} // namespace
+
+AdaptiveLoadClientMain::AdaptiveLoadClientMain(int argc, const char* const* argv,
+                                               AdaptiveLoadController& controller,
+                                               Envoy::Filesystem::Instance& filesystem)
+    : controller_{controller}, filesystem_{filesystem} {
+  TCLAP::CmdLine cmd("Adaptive Load tool that finds the optimal load on the target "
+                     "through a series of Nighthawk Service benchmarks.",
+                     /*delimiter=*/' ', VersionInfo::version());
+
+  TCLAP::ValueArg<std::string> nighthawk_service_address(
+      /*flag_name=*/"", "nighthawk-service-address",
+      "host:port for Nighthawk Service. To enable TLS, set --use-tls.",
+      /*required=*/false, "localhost:8443", "string", cmd);
+  TCLAP::SwitchArg use_tls(
+      /*flag_name=*/"", "use-tls",
+      "Use TLS for the gRPC connection from this program to the Nighthawk Service. Set environment "
+      "variable GRPC_DEFAULT_SSL_ROOTS_FILE_PATH to override the default root certificates.",
+      cmd);
+  TCLAP::ValueArg<std::string> spec_filename(
+      /*flag_name=*/"", "spec-file",
+      "Path to a textproto file describing the adaptive load session "
+      "(nighthawk::adaptive_load::AdaptiveLoadSessionSpec).",
+      /*required=*/true, /*default_value=*/"", "string", cmd);
+  TCLAP::ValueArg<std::string> output_filename(
+      /*flag_name=*/"", "output-file",
+      "Path to write adaptive load session output textproto "
+      "(nighthawk::adaptive_load::AdaptiveLoadSessionOutput).",
+      /*required=*/true, /*default_value=*/"", "string", cmd);
+
+  Nighthawk::Utility::parseCommand(cmd, argc, argv);
+
+  nighthawk_service_address_ = nighthawk_service_address.getValue();
+  use_tls_ = use_tls.getValue();
+  spec_filename_ = spec_filename.getValue();
+  output_filename_ = output_filename.getValue();
+}
+
+uint32_t AdaptiveLoadClientMain::Run() {
```

Comment:
can we file a future bug to enable some dependency injection of the channel here?  in the future, people might want to enable other ways of connecting to the nighthawk service besides the two provided here.  but we don't need to prematurely allow for that case until we have a specific usecase.

Reply:
Filed #626.